[Sema] Fix availability checking in inlinable code

xymus opened the pull request:
Inlinable and `@_alwaysEmitIntoClient` functions can be inlined in clients with a lower OS target version than the framework defining the function. For this reason, the availability in inlinable functions should always be checked by using the introduction OS version specified by the explicit attributes as lower bound and not the minimum deployment version of the framework.

A side-effect of this change is rejecting unavailable overrides to available methods, in a similar way as overrides that are less available than the introduction are rejected. I believe this is the intended behavior as it provided inconsistent diagnostics and surprising behaviors.

rdar://problem/67975153

xymus committed:
[Sema] Check availability in inlinable code using the explicit version

Inlinable functions can be inlined in clients with a lower OS target
version than the framework defining the function. For this reason, the
availability in inlinable functions should always be checked using the
explicit introduction OS version as lower bound and not the minimum
deployment version.

rdar://problem/67975153

xymus committed:
[Sema] Test explicit versions in TypeRefinementContext

xymus committed:
[Sema] Consider unavailable functions as being unreachable

This has the effect of rejecting unavailable overrides to available
methods in a similar way as overrides that are less available than the
introduction are rejected.

xymus committed:
[SIL] Don't print availability for unreachable/unavailable functions

xymus commented:
@swift-ci Please smoke test

xymus commented:
@swift-ci Please test source compatibility

slavapestov commented on include/swift/AST/TypeRefinementContext.h:
```diff
@@ -156,14 +156,17 @@ class TypeRefinementContext {
+
   std::vector<TypeRefinementContext *> Children;
 
   TypeRefinementContext(ASTContext &Ctx, IntroNode Node,
```
Maybe add doc comments for Info / InfoExplicit?

xymus replied:
Done!

xymus committed:
[Sema] Still use the current minimum deployment target as fallback

xymus committed:
[Sema] Prioritize unavailability check for protocol witnesses

The errors on explicit unavailability are more precise than those on
availablity version. Make sure we prioritize the unavailability ones
first.

xymus commented:
@swift-ci Please smoke test

xymus committed:
[Sema] Silence an inconsistent availability error for the stdlib

Don't raise the error availability_decl_more_than_enclosing in the
context is unavailable and if read in a swiftinterface file. This error
was present in the stdlib in 5.3 without side effects so we still want
to accept it.

xymus commented:
@swift-ci Please smoke test

xymus committed:
[Sema] Accept unavailable constructors override in unavailable types

Preserve the old behavior of accepted unavailable override of
constructors as this change would be source breaking.

xymus committed:
[ClangImporter] Import unavailable decls as unavailable

xymus commented:
@swift-ci Please smoke test

xymus commented:
@swift-ci Please smoke test

xymus committed:
[Sema] Create a new TypeRefinementContext only for explicit inlinables

xymus commented:
@swift-ci Please smoke test

xymus commented:
@swift-ci Please smoke test

xymus commented:
@swift-ci Please test source compatibility

shahmishal commented:
Please update the base branch to `main` by Oct 5th otherwise the pull request will be closed automatically. 

* How to change the base branch: ([Link](https://docs.github.com/en/github/collaborating-with-issues-and-pull-requests/changing-the-base-branch-of-a-pull-request))
* More detail about the branch update: ([Link](https://forums.swift.org/t/updating-branch-names/40412/))

xymus commented:
@swift-ci Please test source compatibility

xymus committed:
[Sema] Use the deployment target as minimum OS version in swiftinterfaces

Preserve the previous behavior when parsing swiftinterfaces as the Swift
stdlib for Swift 5.3 and a few other swiftinterfaces in Xcode 12.0 had
issues that would now be reported as errors.

xymus commented:
@swift-ci Please smoke test

xymus commented:
@swift-ci Please test source compatibility

xymus commented:
@swift-ci Please smoke test macOS platform

xymus commented:
@swift-ci Please smoke test macOS platform

xymus commented:
@swift-ci Please test source compatibility

xymus commented:
@swift-ci Please smoke test

xymus commented:
@swift-ci Please test source compatibility

xymus commented:
@swift-ci Please smoke test macOS platform

xymus commented:
@swift-ci Please smoke test Linux platform

xymus commented:
@swift-ci Please test source compatibility

xymus commented:
@swift-ci Please smoke test

xymus commented:
@swift-ci Please test source compatibility

xymus commented:
@swift-ci Please smoke test

brentdax commented:
Restarting doomed Linux tests.

@swift-ci please smoke test Linux platform

xymus commented:
@swift-ci Please test source compatibility

xymus commented:
@swift-ci Please test source compatibility

xymus committed:
[Sema] Report non-constructor unavailable overrides as warnings

Report unavailable overrides as a warning to avoid breaking sources.

xymus commented:
@swift-ci Please smoke test

xymus commented:
@swift-ci Please test source compatibility
